Accurately find all baseline directories

gsnedders opened the pull request:
#### 008f6b91afed5564eabc5dc02f2b7880b614c852
<pre>
Accurately find all baseline directories
<a href="https://bugs.webkit.org/show_bug.cgi?id=248340">https://bugs.webkit.org/show_bug.cgi?id=248340</a>

Reviewed by Jonathan Bedard.

* Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_finder_legacy_unittest.py:
(LayoutTestFinderTests.test_include_other_platforms_fallback):
* Tools/Scripts/webkitpy/port/apple.py:
(ApplePort.all_baseline_search_paths):
* Tools/Scripts/webkitpy/port/base.py:
(Port.all_baseline_search_paths):
(Port._tests_for_other_platforms):

Canonical link: <a href="https://commits.webkit.org/257480@main">https://commits.webkit.org/257480@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4c0846264e0a27bad1de0362c2a4ae7e465c2599

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99017 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8223 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32169 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108423 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/168674 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/102979 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/8776 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85560 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91539 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106380 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/104738 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/6642 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/90225 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33666 "Passed tests") | 
| [❌ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/102553 "Found 1 webkitpy python2 test failure: webkitpy.port.server_process_unittest.TestServerProcess.test_broken_pipe") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/88464 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21570 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/76520 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2124 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23086 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2027 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/45477 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5147 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6994 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42561 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/3442 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

JonWBedard commented on Tools/Scripts/webkitpy/port/apple.py:
```diff
@@ -113,6 +113,26 @@ def _generate_all_test_configurations(self):
+            self.get_option("additional_platform_directory", [])
+            + self._compare_baseline()
+            + (
+                list(self._filesystem.glob(self._apple_baseline_path("*")))
```
I would need to do some local spelunking to verify this concern, but our Apple specific baselines should be woven into our WebKit ones. So it should be something like:
```
apple-platform/ios-simulator
LayoutTests/platform/ios-simulator
apple-platform/ios
LayoutTests/platform/ios
```
instead of:
```
apple-platform/ios-simulator
apple-platform/ios
LayoutTests/platform/ios-simulator
LayoutTests/platform/ios
```
(which is what I'm afraid this change will do)

gsnedders replied:
You are, of course, technically correct. I guess given it's trying to provide all possible baseline directories it should really return a `set` rather than a `list`, which perhaps makes it clearer that there is no meaningful ordering here.

JonWBedard replied:
Ah! That makes sense, I always expect our baseline search path to be ordered, returning a `set` would be fine, even just a comment that says "hey, order doesn't matter here, even though it usually does for this list"

gsnedders replied:
Right—it's not an unreasonable expectation, it's just we have no reasonable ordering for "does win7 or gtk come first?".

gsnedders committed:
Accurately find all baseline directories
https://bugs.webkit.org/show_bug.cgi?id=248340

Reviewed by Jonathan Bedard.

* Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_finder_legacy_unittest.py:
(LayoutTestFinderTests.test_include_other_platforms_fallback):
* Tools/Scripts/webkitpy/port/apple.py:
(ApplePort.all_baseline_search_paths):
* Tools/Scripts/webkitpy/port/base.py:
(Port.all_baseline_search_paths):
(Port._tests_for_other_platforms):

Canonical link: https://commits.webkit.org/257480@main
